### PR TITLE
feat (mapfilter) - Add mapFilter operator + docs

### PR DESCRIPTION
**Pull request by dmorosinotto**
I added the mapFilter operator proposed on #34 
after that I'll create a separate PR for debug operator.
The commit also added the mapfilter.md for docs, don't found the astro config for link it, maybe now is it all automated!?

### CODE REVIEW

File: docs/src/content/docs/utilities/Operators/mapfilter.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
+---
+title: mapFilter
+description: An RxJS operator that allow to apply a trasform function to each value of the observable in (same as map), but with the ability to skip (filter out) some values if the function explict return undefined or simply doesn't return anything for same code-path (implict return undefined).
```
Comment (ajitzero) [inline docs/src/content/docs/utilities/Operators/mapfilter.md:3]:
```suggestion
description: An RxJS operator that allows applying a transform function to each value of the observable in (same as map), but with the ability to skip (filter out) some values if the function explicit return undefined or simply doesn't return anything for same code-path (implicit return undefined).
```
Minor typos/grammar

File: docs/src/content/docs/utilities/Operators/mapfilter.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+## Import
+
+```typescript
+import { mapFilter } from 'ngxtension/mapfilter';
+```
+
+## Usage
+
+You can use it as a normal map operator, but with the ability to skip some values: returning undefined (explict or implict).
```
Comment (ajitzero) [inline docs/src/content/docs/utilities/Operators/mapfilter.md:14]:
```suggestion
You can use it as a normal map operator, but with the ability to skip some values: returning undefined (explicit or implicit).
```

File: libs/ngxtension/mapfilter/project.json [hunk 3]
```diff
@@ -0,0 +1,33 @@
+{
+	"name": "ngxtension/mapfilter",
```
Comment (ajitzero) [inline libs/ngxtension/mapfilter/project.json:2]: nit: Not sure about the naming scheme but should the file be `ngxtension/map-filter` instead (hyphen in middle?)

Comment (dmorosinotto) [reply libs/ngxtension/mapfilter/project.json:2]: I don't know if there is an enforced/preferred naming convention for the files...

**Comment by dmorosinotto** [comment]
@nartc about the failing CI I need to write the @tsignore inside the TEST to prove that the mapFunction correctly handle the case when someone don't produce/return any value for some code-path.
Locally the test pass, but not in CI can you please help me or suggest a valid workaround? Thanks

File: docs/src/content/docs/utilities/Operators/mapfilter.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
+---
+title: mapFilter
+sdescription: An RxJS operator that allows applying a transform function to each value of the observable in (same as map), but with the ability to skip (filter out) some values if the function explicit return undefined or simply doesn't return anything for same code-path (implicit return undefined).
```
Comment (ajitzero) [inline docs/src/content/docs/utilities/Operators/mapfilter.md:3]:
```suggestion
description: An RxJS operator that allows applying a transform function to each value of the observable in (same as map), but with the ability to skip (filter out) some values if the function explicitly returns undefined or simply doesn't return anything for same code-path (implicit return undefined).
```
One more

File: libs/ngxtension/mapfilter/src/mapfilter.ts [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import { type Observable } from 'rxjs';
+import { filter, map } from 'rxjs/operators';
+
+export function mapFilter<T, R>(fnTrasformSkipUndefined: (value: T) => R) {
+	return function (source: Observable<T>): Observable<Exclude<R, undefined>> {
+		return source.pipe(
+			map(fnTrasformSkipUndefined),
+			filter((value) => value !== undefined)
```
Comment (tomalaforge) [inline libs/ngxtension/mapfilter/src/mapfilter.ts:8]: you can use `filterNil()`here, this way I don't think you need to cast you observable at the end.

File: libs/ngxtension/mapfilter/src/mapfilter.spec.ts [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+			})
+		);
+
+		let out = '>';
+		out$.subscribe((s) => {
+			//s is a string
+			out += s + '-';
+		});
+		out += '|';
+		expect(out).toEqual('>2-6-|');
```
Comment (tomalaforge) [inline libs/ngxtension/mapfilter/src/mapfilter.spec.ts:20]: I don't get your test. Why are you not simply testing the returned value ? It makes your test very hard to understand. 
We could add observableSpy to eaiser test observable. What do you thing @nartc ?

File: libs/ngxtension/mapfilter/project.json (hunk 3 above)
Comment (dmorosinotto) [reply libs/ngxtension/mapfilter/project.json:2]: Ok @nartc I'll rename all files in next commit
PS: Do you have any suggestion about the @ts-ignore problem for CI?

File: libs/ngxtension/mapfilter/src/mapfilter.ts (hunk 5 above)
Comment (dmorosinotto) [reply libs/ngxtension/mapfilter/src/mapfilter.ts:8]: @tomalaforge sorry I can't use your filterNil implementation because I want to filter ONLY undefined, null is a valid return value that must propagate through the filter, and so even the explicit cast at the end is needed.

File: libs/ngxtension/mapfilter/src/mapfilter.spec.ts (hunk 6 above)
Comment (dmorosinotto) [reply libs/ngxtension/mapfilter/src/mapfilter.spec.ts:20]: @tomalaforge I'm not an test expert expecially for RxJS, don't know how to use/rewrite it using observableSpy, in either case all my test passes! 
And my code is simply composing an out string that "mimic" the "marble diagram" of the out$, because I need to test the result checking the stream of events that occurs, and the single value that can pass through mapfilter (odd numbers -> 2x) and which one are filtered (even number)

Comment (dmorosinotto) [reply libs/ngxtension/mapfilter/src/mapfilter.spec.ts:20]: Even if I use for semplicity a "sync" observable alle the value 1, 2, 3 are emitted as single number, notice I use in$= from([1,42,3]) that will produce an Observable with a "marble" representation like  this `>1-42-3-|` where 
`>` represents the start of the stream
`-` represents the time that sepatete one next from the subsequent
`|` is the complete event (end of the stream) 
`X` here not used, represents and error emission (that maybe end the stream)

Hope this will clarify even my test cases

Comment (tomalaforge) [reply libs/ngxtension/mapfilter/src/mapfilter.spec.ts]: Yes I get the 'look like' marble test. But real marble test are hard to understand, so similar one are even harder. It's hard to get why you test it like that. 

My propose solution:
```ts
	it('given an observable >1-42-3| and a mapping function that double ONLY the odds value, then result is an observable of >"2"--"6"-| the intial even value (42) is not mapped and so filtered out', fakeAsync(() => {
		const in$ = of(1, 42, 3);
		const out$ = in$.pipe(
			mapFilter((n) => {
				if (n % 2) return String(n * 2);
				else return undefined; // explict return undefined to skipout (filter) some value from the out observable
			})
		);

		// let out = '>';
		out$.pipe(toArray()).subscribe((s) => {
			//s is a string
			expect(s).toEqual(['2', '6']);
		});
		tick();
		// out += '|';
		// expect(out).toEqual('>2-6-|');
	}));
```

or 
```ts
	it('given an observable >1-42-3| and a mapping function that double ONLY the odds value, then result is an observable of >"2"--"6"-| the intial even value (42) is not mapped and so filtered out', (done) => {
		const in$ = of(1, 42, 3);
		const out$ = in$.pipe(
			mapFilter((n) => {
				if (n % 2) return String(n * 2);
				else return undefined; // explict return undefined to skipout (filter) some value from the out observable
			})
		);

		// let out = '>';
		out$.pipe(toArray()).subscribe((s) => {
			//s is a string
			expect(s).toEqual(['2', '6']);
			done();
		});
		// out += '|';
		// expect(out).toEqual('>2-6-|');
	});
```

File: libs/ngxtension/mapfilter/src/mapfilter.ts (hunk 5 above)
Comment (dmorosinotto) [reply libs/ngxtension/mapfilter/src/mapfilter.ts]: Because this is **MINE** desired implementation, and if you read the docs description: An RxJS operator that allows applying a transform function to each value of the observable in (same as map), but with the ability to skip (filter out) some values if the function explicitly returns undefined or simply doesn't return anything for same code-path (implicit return undefined).

I clarify that only undefined is filter out never spoken about null.
And the motivation is really simple even if "null+undefined" is sometimes called the "2 billion$ mistake" for me it's really clear that in JS undefined is a way of the JS Engine/runtime to express missing value, null is our (dev-world) to express missing value.

The intent of mapFilter is just to give to the dev-user a way to write an imperative trasform that choose which value to return (even null if they want, it's always a valid value in JS/TS) + skips some value emit to the down streams, in 2 simple ways: explicit return undefined, or either write a transform function that for some "code-path" doesn't return anything --> so the function for JS implicit return undefined (AND NOT null) and this is why I choose to filter out ONLY undefined.

If someone still want to filter out even null, can simply done `in$.pipe(mapFilter(...), filterNil())`

File: libs/ngxtension/mapfilter/src/mapfilter.spec.ts (hunk 6 above)
Comment (dmorosinotto) [reply libs/ngxtension/mapfilter/src/mapfilter.spec.ts]: Thanks for your suggested Test code, I've never known/used `toArray()` operator, quite un-usual one, but in this case can simplify the test expect. 
If you don't mind rewrite them I prefer to keep the "marble" representation...
PS: with **of**`([1,42,3])` the test will fail, you must use **from**`([1,42,3])`;

File: libs/ngxtension/mapfilter/src/mapfilter.ts (hunk 5 above)
Comment (tomalaforge) [reply libs/ngxtension/mapfilter/src/mapfilter.ts]: Never wanted to upset you, that's why I added "question" in front. I was just asking a question. I'm out of this discussion. I leave @nartc or @eneajaho handle this.

Comment (dmorosinotto) [reply libs/ngxtension/mapfilter/src/mapfilter.ts]: @tomalaforge sorry for the tone, maybe too rude  
I didn't got your question intent initially... sorry 😞! 
Hope you can understand the motivation way I want to keep null, and filter only undefined.
Have a nice day/night

File: libs/ngxtension/mapfilter/src/mapfilter.spec.ts (hunk 6 above)
Comment (nartc) [reply libs/ngxtension/mapfilter/src/mapfilter.spec.ts]: I second the `observer-spy` approach. The tests work but they're hard to parse what's going on.

File: libs/ngxtension/mapfilter/src/mapfilter.ts (hunk 5 above)
Comment (nartc) [reply libs/ngxtension/mapfilter/src/mapfilter.ts]: I agree with @tomalaforge. In my opinion, the value proposition of `mapFilter` comes from the narrower type that `filter((value) => value !== undefined` provides via `Exclude<R, undefined>`. 

`mapFilter` as-is is too limiting. If we say `mapFilter`, I'd expect to be able to pass in a `filterFn` as well as the `transformFn`. 

My proposal is one of the following:
1. rename to `mapFilterUndefined` or `mapSkipUndefined`
2. pivot to a different operator `filterUndefined` or `skipUndefined` to only provide the type-narrow feature
```ts
$.pipe(
    map(mapFn),
    filterUndefined() // same as filter(val => val !== undefined) but it is more strongly typed
)
```

File: libs/ngxtension/mapfilter/project.json (hunk 3 above)
Comment (nartc) [reply libs/ngxtension/mapfilter/project.json]: You can add `noImplicitReturns: false` to `tsconfig.spec.json`

File: libs/ngxtension/mapfilter/src/mapfilter.spec.ts (hunk 6 above)
Comment (dmorosinotto) [reply libs/ngxtension/mapfilter/src/mapfilter.spec.ts]: Is this [observer-spy](https://www.npmjs.com/package/@hirez_io/observer-spy) you are referring?
Do I can install it? And have to update the jest.config to handle auto-unsubscribe?!

Please @nartc can you suggest me how to solve the problem with CI failing because I used the @ts-ignore to write the test for empty code-path that doesn’t return value, or I simply remove the test before committing…

File: libs/ngxtension/mapfilter/src/mapfilter.ts (hunk 5 above)
Comment (dmorosinotto) [reply libs/ngxtension/mapfilter/src/mapfilter.ts]: I don't think that people need a new operator to simply pipe(map, filter) if they need to pass in all the fnTrasform and fnFilter, so I'll go fr the 1) chooise to rename mapFilter -> `mapSkipUndefined` with keep it consistent with current implementation.

When I'll have time I'll rename/rewrite all the files , docs, and tests...

@nartc about testing with `observer-spy` please can you create a commit on this branch, or in main to introduce the new dependency and config for jest or whatever it'll be necessary in the NX workspace to make it works, next I'll rewrite the tests, Thanks

**Comment by nartc** [comment]
We'll revisit the PR after `observer-spy` is in. At the moment, can you fix the rebase since your PR contains code that's already in `main`?

**Comment by dmorosinotto** [comment]
@nartc I'll close this PR and tomorrow will redone all works from ground up in a new PR correct naming `mapSkipUndefined`, and simplifying test using `toArray()` as suggested by @tomalaforge hoping this will make everyone happy and then will be approved.

I need to go sleep tonight, tomorrow I'll done it, promised!
